src/TypeInference/PreludeTypes.ml

```diff
@@ -0,0 +1,72 @@
+(* This file is part of DBL, released under MIT license.
+ * See LICENSE for details.
+ *)
+
+(** Functions allowing the use of some of the types defined in the Prelude. *)
+
+open Common
+
+let mk_Option ~env ~pos tp_arg =
+  match Env.lookup_tvar env (S.NPName "Option") with
+  | None -> Error.fatal (Error.unbound_type_var ~pos (S.NPName "Option"))
+  | Some tp ->
+      if
+        not
+          (Unification.unify_kind (T.Type.kind tp)
+             (T.Kind.k_arrow T.Kind.k_type T.Kind.k_type))
+      then
+        Error.fatal
+          (Error.kind_mismatch ~pos (T.Type.kind tp)
+             (T.Kind.k_arrow T.Kind.k_type T.Kind.k_type))
+      else T.Type.t_app tp tp_arg
+
+let option_to_whnf ~env ~pos option_tp =
+  match T.Type.whnf option_tp with
+  | Whnf_Neutral (NH_Var x, tp :: []) -> (tp, x)
+  | _ -> Error.fatal (Error.invalid_whnf_form ~pos ~env option_tp)
+
+let extr_arg_tp ~env ~pos option_tp =
+  let tp, _ = option_to_whnf ~env ~pos option_tp in
+  tp
+
+let mk_Some ~env ~pos tp_arg expr_arg : T.expr =
+  let make data pos = { T.data; T.pos } in
+  let tp' = mk_Option ~env ~pos tp_arg in
+  let _, x = option_to_whnf ~env ~pos tp' in
+  match Env.lookup_adt env x with
+  | Some adt_info -> (
+      let { Module.adt_proof; adt_args; adt_ctors; adt_type } = adt_info in
+      match
+        List.find_index
+          (fun (ctor_decl : T.ctor_decl) -> String.compare ctor_decl.ctor_name "Some" == 0)
+          adt_ctors
+      with
+      | Some idx ->
+          make
+            (T.ECtor
+               ( make (T.ETApp (adt_proof, tp_arg)) Position.nowhere,
+                 idx,
+                 [],
+                 [ expr_arg ] ))
```

Comment:
Do we know, that `Some` constructor expects one parameter? Should it be checked? I know, such errors should never happen for casual user, but ...

Reply:
I tried to implement it, checking correct type of `Some` and `None` constructor. The code is too complicated for a thing that should never happen, so I decided to not address this issue. Internal type error is good enough.